cached.go

```diff
@@ -0,0 +1,69 @@
+package ydb
+
+import (
+	"crypto/sha256"
+	"crypto/x509"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/ydb-platform/ydb-go-sdk/v3/internal/xerrors"
+)
+
+var DisableFileCache = false
+
+var fileCache sync.Map
+
+// readFileCached is a cached version of os.ReadFile. Cache key is
+//  filepath.Clean(file)
+func readFileCached(file string) ([]byte, error) {
+	file = filepath.Clean(file)
+	if !DisableFileCache {
+		value, exists := fileCache.Load(file)
+		if exists {
+			bytes, ok := value.([]byte)
+			if !ok {
+				panic("unknown file cache type")
+			}
+			return bytes, nil
+		}
+	}
+	bytes, err := os.ReadFile(file)
+	if err != nil {
+		return nil, xerrors.WithStackTrace(err)
+	}
+	if !DisableFileCache {
+		fileCache.Store(file, bytes)
+	}
+	return bytes, nil
+}
+
+var DisableCertificateCache = false
+
+var certificateCache sync.Map
+
+// parseCertificateCached is a cached version of x509.ParseCertificate. Cache key is
+//  sha256.Sum224(der)
+func parseCertificateCached(der []byte) (*x509.Certificate, error) {
+	// using Sum224 for cert data similarly to tls package
+	var sum [sha256.Size224]byte
+	if !DisableCertificateCache {
+		sum = sha256.Sum224(der)
```

Comment:
What do you think about key as string(der)?
Any crypto hashing algorithms:
- uses a lot of CPU
- have collisions

Reply:
May be search by looooooong string have the same CPU usage...

Reply:
So, a quick [bench](https://gist.github.com/xopoww/30f2ea0d88580c2d1323dfa392fd0424) shows that using `string(der)` is faster, but memory usage in this case is worse. Example bench result:
```
BenchmarkKeys/0/sha224/store____
BenchmarkKeys/0/sha224/store____-8                181789              5886 ns/op             392 B/op          8 allocs/op
BenchmarkKeys/0/sha224/load_hit_
BenchmarkKeys/0/sha224/load_hit_-8                203478              5412 ns/op              32 B/op          1 allocs/op
BenchmarkKeys/0/sha224/load_miss
BenchmarkKeys/0/sha224/load_miss-8                198758              5438 ns/op              32 B/op          1 allocs/op
BenchmarkKeys/0/string
BenchmarkKeys/0/string/store____
BenchmarkKeys/0/string/store____-8               1245193              1252 ns/op            2424 B/op          9 allocs/op
BenchmarkKeys/0/string/load_hit_
BenchmarkKeys/0/string/load_hit_-8               2395186               494.1 ns/op          2064 B/op          2 allocs/op
BenchmarkKeys/0/string/load_miss
BenchmarkKeys/0/string/load_miss-8               2732768               450.2 ns/op          2064 B/op          2 allocs/op
```

Another idea might be to somehow exploit binary format of x509 certificate without actually parsing it, using some sub-slice of data (which is guaranteed to be unique between certs) as key. But it seems to unstable to use.